README.md
# Amazon Ads API SDK

![GitHub](https://img.shields.io/github/license/whitebox-co/amazon-ads-api)
![GitHub](https://img.shields.io/github/package-json/v/whitebox-co/amazon-ads-api)
![GitHub Workflow Status](https://img.shields.io/github/workflow/status/whitebox-co/amazon-ads-api/Release?label=main)
![GitHub Workflow Status](https://img.shields.io/github/workflow/status/whitebox-co/amazon-ads-api/Develop?label=develop)
![GitHub Workflow Status](https://img.shields.io/github/workflow/status/whitebox-co/amazon-ads-api/Generate%20API?label=api-generator)
![GitHub Workflow Status](https://img.shields.io/snyk/vulnerabilities/github/whitebox-co/amazon-ads-api?label=snyk%20vulnerbilities)
![GitHub Workflow Status](https://img.shields.io/maintenance/yes/2022)

A fully typed and auto-generated [TypeScript](https://www.typescriptlang.org) library for the
[Amazon Ads API](https://advertising.amazon.com/API/docs/en-us/info/api-overview).

This library automatically generates typescript types, classes, interfaces, integration tests, and UI documentation
based directly off of the Amazon Ads OpenAPI specification(s).

## Why?

TLDR: It did not exist, and we needed it!

-   Amazon does not provide a Typescript, NodeJs, or Javascript library to access their endpoints.
-   Existing implementations, the few that there are, did not include full the full Amazon Ads REST implementation,
    were not auto-generated based on the schemas, or did not meet our criteria for using open-source software.

## How it works

-   An [OpenAPI Schema](https://spec.openapis.org/oas/latest.html) is downloaded from Amazon Ads for each section of the
    Amazon Ads API.
-   API Classes are generated using the amazing [openapi-generator](https://github.com/OpenAPITools/openapi-generator)
    against each schema.
-   Integration Test stubs are auto-generated for every generated API class and function.
-   Documentation is auto-generated and updated based on the latest OpenAPI schema changes.

## Installation

Packages are released on both [Github](https://github.com/whitebox-co/amazon-ads-api/packages) and
[NPM](https://www.npmjs.com/package/@whitebox-co/amazon-ads-api).

To install the library:

```sh
npm install @whitebox-co/amazon-ads-api
```

For development, assuming all of the prerequisites have been satisfied.

```sh
npm install
```

## Usage

There are two distinct parts of this SDK.

1. **Models** - The Models contain interfaces and classes that define the individual API endpoint definitions and
   extend a base API class that ultimately implements Axios calls to the remote endpoint. Sse these classes and models
   directly when you need to manage the configuration and authorizations yourself.
2. **Clients** - The Clients are classes that group multiple APIs into one named group. These API's map directly to the
   Amazon OpenAPI Swagger definitions. They extend multiple APIs and group all functions of all APIs into a single
   exported class. These are the main entry points into using the SDK to connect to the remote endpoints.

To properly authorize against each endpoint, the Amazon Ads API requires passing both access tokens and refresh tokens.
Our SDK handles this transparently to the end-user by exposing a mechanism to manage the cache and to refresh the
access tokens. To take advantage of this, the user must use the `getConfiguredApi` function to instantiate an
individual client.

### Preferred Usage

Use this approach when you only need to use a single client. If multiple clients are necessary
it is better to use the AmazonAds Class.

```typescript
import amazonAdsApi, { AttributionClient, AmazonAdsApiConfiguration, setConfiguration } from '@whitebox-co/amazon-ads-api';

// The optional configuration
const configuration: AmazonAdsApiConfiguration = {
    throttling: {
        reservoir: 100,
        reservoirRefreshAmount: 100,
        reservoirRefreshInterval: 60 * 1000,
        maxConcurrent: 5,
        minTime: 333,
    },
	retries: {
		count: 3,    // defaults to 3
        refreshTime: 5000, // in ms. defaults to 5000
        maxWaitTime: 5000, // in ms. defaults to 5000
		retryCallback?: (jobId: string) => { console.log(`Retrying job ${jobId}`); }  // optional
	},
    jobOptions: {
        expiration: 3600 * 1000,
    },
	envoyProxyRateLimitUrl: 'http://localhost:8080' // optional (see https://github.com/envoyproxy/ratelimit)
};

// set the configuration options (totally optional, defaults are set above)
setConfiguration(configuration);

/**
 * Init the amazon ads, attribution client, with client credentials and get back a fully
 * configured instance of the client with token caching built-in.
 *
 * Any subsequent call using the same credentials but a different client
 * will use cached credentials until they expire.
 */
const attributionClient = await amazonAdsApi.getConfiguredApi(AttributionClient, {
    clientId
    clientSecret,
    profileId,
    refreshToken,
});

/**
 * Example of a GET call to the attribution API.
 */
const response = await attributionClient.getAdvertisersByProfile({
    amazonAdvertisingAPIClientId: '',
    amazonAdvertisingAPIScope: ''
});
```

### Class-Based Alternative

This approach is better to use when you need multiple client's and only want to have to pass in and
configure credentials one time.

```typescript
import { AmazonAds, AttributionClient } from '@whitebox-co/amazon-ads-api';

// The optional configuration
const configuration: AmazonAdsApiConfiguration = {
    throttling: {
        reservoir: 100,
        reservoirRefreshAmount: 100,
        reservoirRefreshInterval: 60 * 1000,
        maxConcurrent: 5,
        minTime: 333,
    },
	retries: {
		count: 3,    // defaults to 3
        refreshTime: 5000, // in ms. defaults to 5000
        maxWaitTime: 5000, // in ms. defaults to 5000
		retryCallback?: (jobId: string) => { console.log(`Retrying job ${jobId}`); }  // optional
	},
    jobOptions: {
        expiration: 3600 * 1000,
    },
	envoyProxyRateLimitUrl: 'localhost:8080' // optional (see https://github.com/envoyproxy/ratelimit)
};

// The user provided api credentials
const apiCredentials: AmazonAdvertisingAPICredentials = {
    clientId
    clientSecret,
    profileId,
    refreshToken,
}

/**
 * Init a new instance of the AmazonAds class.
 *
 * This has the advantage of only having to pass in credentials once.
 */
const amazonAdsApi = new AmazonAds(apiCredentials, configuration);

/**
 *  Retrieves a fully authorized instance of the attribution client.
 */
const attributionClient = await amazonAdsApi.getConfiguredApi(AttributionClient);

/**
 * Example of a GET call to the attribution API.
 */
const response = await attributionClient.getAdvertisersByProfile({
    amazonAdvertisingAPIClientId: '',
    amazonAdvertisingAPIScope: ''
});
```

## Docs

There are three different types of documentation included for this repo.

1. **Redoc** - Used as an alternative to visiting the amazon ads API site.
2. **SwaggerUI** - Used as an alternative to Redoc.
3. **TypeDoc** - Typescript type and class documentation of the main library.

Redoc and SwaggerUI documentation included as an alternative to visiting the Amazon Ads developer website.

### Viewing Docs

To quickly view the documentation, we have packaged [http-server](https://github.com/http-party/http-server) as a
dev dependency. Issue any of the following commands to build and view any docs.

```sh
npm run docs # or `npm run docs:typedoc`
npm run docs:redoc
npm run docs:swaggerui
```

**Note:** `http-server` will try and serve the docs on port `8080` and will auto-launch your browser to the
specified documentation.

## Development

### Prerequisites

-   [Java Runtime](https://www.java.com/en/download/manual.jsp) - Necessary to run the `openapi-generator`
-   [Node](https://nodejs.org/en/) - To install and build the library.
-   [Amazon Ads API Experience](https://advertising.amazon.com/API/docs/en-us/info/api-overview) - Helpful to have
    experience with the Amazon Ads REST API.

### Commands

```sh
npm run build       # build the project (output to ./lib)
npm run dev         # run ts-node-dev to watch and rebuild project while in development.
npm run lint        # runs eslint linter
npm test -- -i/-u -t "<DESCRIBE_OR_IT_NAME>" # Example: npm test -- -i -t "Attribution"
```

### API Generation

API Generation is handled automatically by Github actions. We perform a task daily to check if any of the Amazon Ads
APIs have been updated and, if it has, we perform the schema downloads and subsequent API Generation. We then
automatically generate the new docs and, a new version of the library is built and deployed to npm.

To manually perform those steps, see the sections below.

#### Manual Schema Download

Amazon exposes OpenApi 3.0 schemas _for most_ of their APIs. To pull down all of the schemas, run the following:

```sh
npm run download-schemas
```

If changes have occurred to any schema, the new updates will overwrite the existing schemas in the `./docs/schemas`
directory.

#### Manual Model Generation

Once you have the schemas downloaded and then you can run the model generation by running the following:

```sh
npm run generate-models
```

Which will then start the generation and processing of all of the schemas and eventually finish with all models
created in the `src/apis/models` directory.

#### Manual Client Generation

Once you have the models generated, you can run the client generation by running the following:

```sh
npm run generate-clients
```

Which will start the client generation and eventually finish with all clients created in the `src/apis/clients`
directory.

#### Manual API Generation

If you wish to do all the above generation steps in one go, run the following:

```sh
npm run generate-apis
```

Which will download schemas, generate models, generate clients, generate tests, and output everything to their
respective folders. Github actions are currently set up to run this process on a daily basis.

## Token Authorization and Caching

Tokens are retrieved from amazon during the LWA (Login With Amazon) authorization process and then get cached until
they expire. When the access token expires, the library issues another call to the amazon refresh endpoint to retrieve
a new access token.

Any time you request a new API from the amazonAdsApi using the same clientID, the cached credential token
will be used to authorize the request.

## Issues

Please report any issues you find in the Github issues section. We will do our best to address any issue in a timely
manner.

### Known Issues

-   Some of the Sponsored Brands API models have `uNKNOWNBASETYPE?: any` in places that the open-api-generator can not
    properly figure out how to handle the complexities of `allOf/anyOf` inheritance. This is an issue with
    open-api-generator and, we have not yet found a workaround. For now, the best solution is to directly consult the
    OpenApi schema when trying to determine which request params to use for these API calls.
-   Amazon Ads APIs do not always provide schemas for their APIs. In those scenarios we have created our own based
    of of their text based documentation. You can see which those are by referencing the `constants.ts` file and
    looking for the local references of the schemas.
-   Amazon does not always update an existing schema per ads category and in some instances they simply make a new
    schema altogether and just add it to an existing category. For instance, see [Sponsored Brands Campaign](https://advertising.amazon.com/API/docs/en-us/sponsored-brands/3-0/openapi#/Campaigns) vs [Sponsored Brands Budget Rules](https://advertising.amazon.com/API/docs/en-us/sponsored-brands/3-0/openapi/prod#/BudgetRules).
    You will notice that both of these have separate schema definitions. In these scenarios we will append "Extra" to
    the schema name to signify the differences between these. In the future, the hope is that Amazon will concatenate
    these into an API under a new version.

## Acknowledgements

Thanks to the following organizations for their help and inspiration.

-   [ScaleLeap](https://github.com/ScaleLeap) - For their work on auto-generation of the Selling Partner API, which is
    our initial inspiration for this project.
-   [Whitebox](https://whitebox.com/) - For their commitment to providing open-source contributions. Without their
    support, this project would not exist.

## Contributing

Feel free to open PRs. [Whitebox](https://whitebox.com/) is currently using this in our production code, and we will
evaluate changes on a case-by-case basis.

See the [Contributing](CONTRIBUTING.md) documentation for further details.

## Code of Conduct

This project has adopted the [Contributor Covenant Code of Conduct](https://www.contributor-covenant.org/).

See the [Code Of Conduct](CODE_OF_CONDUCT.md) documentation for further details.

## License

![GitHub](https://img.shields.io/github/license/whitebox-co/amazon-ads-api)

A copy of the license is provided in the root directory under [LICENSE](LICENSE)


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [2.0.0](https://github.com/whitebox-co/amazon-ads-api/compare/v1.6.0...v2.0.0) (2022-12-17)

### Bug Fixes

-   adding skipLibCheck to fix failing lib build ([febe94a](https://github.com/whitebox-co/amazon-ads-api/commit/febe94a4494b137199793e150aab95b3c7818462))
-   running npm audit fix to update package-lock issues ([77ea22b](https://github.com/whitebox-co/amazon-ads-api/commit/77ea22b2ed4220063dfe7130df9e78a07fe07315))

-   feat(api)!: generating new code based on latest schema changes from amazon ([120fb93](https://github.com/whitebox-co/amazon-ads-api/commit/120fb93415b68754164e59be3746150e46a76b13))

### BREAKING CHANGES

-   InlineResponse generation and schema changes

# [1.6.0](https://github.com/whitebox-co/amazon-ads-api/compare/v1.5.2...v1.6.0) (2022-03-30)

### Features

-   adding portfolios api and integrating new amazon changes ([24fb1d7](https://github.com/whitebox-co/amazon-ads-api/commit/24fb1d71600fa0e657ab0ba512cd30ff92952a37))

## [1.5.2](https://github.com/whitebox-co/amazon-ads-api/compare/v1.5.1...v1.5.2) (2022-03-22)

### Bug Fixes

-   increasing timeouts as large report transfers exceed the 60s threshold ([4ecb0b6](https://github.com/whitebox-co/amazon-ads-api/commit/4ecb0b694afbdfb5f6a939a6148bd8531daabfc6))

## [1.5.1](https://github.com/whitebox-co/amazon-ads-api/compare/v1.5.0...v1.5.1) (2022-03-22)

### Bug Fixes

-   make axiosInstance configurable and increase defaults ([5eefdfa](https://github.com/whitebox-co/amazon-ads-api/commit/5eefdfaf29fa054b406d0d59078299bb3c97169d))

# [1.5.0](https://github.com/whitebox-co/amazon-ads-api/compare/v1.4.2...v1.5.0) (2022-03-22)

### Features

-   adding agentkeepalive for req agents, adding timeouts and maxRedirects to axios instance ([def7eff](https://github.com/whitebox-co/amazon-ads-api/commit/def7effeb0d4fb2c0280ef59092c1a1e234b0971))
-   integration of latest amazon updates ([0fb0482](https://github.com/whitebox-co/amazon-ads-api/commit/0fb04826600d6ccabcd01df6f59b241e5dfe6862))

## [1.4.2](https://github.com/whitebox-co/amazon-ads-api/compare/v1.4.1...v1.4.2) (2022-03-14)

### Bug Fixes

-   retry refresh token logic was not properly awaited resulting in unhandled exceptions ([c33d31a](https://github.com/whitebox-co/amazon-ads-api/commit/c33d31a48df67c0223e9642b52e7a97ad98e8d05))

## [1.4.1](https://github.com/whitebox-co/amazon-ads-api/compare/v1.4.0...v1.4.1) (2022-03-14)

### Bug Fixes

-   removing limiter id to avoid id collisions, limiter will assign its own ids ([0828bc5](https://github.com/whitebox-co/amazon-ads-api/commit/0828bc5ef2ade3e7b093a7bbc3d7d3054947359e))

# [1.4.0](https://github.com/whitebox-co/amazon-ads-api/compare/v1.3.0...v1.4.0) (2022-03-14)

### Bug Fixes

-   **deps:** bump axios from 0.25.0 to 0.26.1 ([#87](https://github.com/whitebox-co/amazon-ads-api/issues/87)) ([86694da](https://github.com/whitebox-co/amazon-ads-api/commit/86694da162795f9a9dcd9a57d5b5316be34087dd))

### Features

-   auto generated updates based on latest amazon schema changes ([2477264](https://github.com/whitebox-co/amazon-ads-api/commit/24772642a69d92595ad7587e5601e54c9eec9691))
-   configuration and adsapi constructor now take a bottleneck limiter as an optional param ([9a2f127](https://github.com/whitebox-co/amazon-ads-api/commit/9a2f127b7ad66bd5ac454030ea6e2c1222d7c32c))
-   integrating latest amazon schema changes, adding ability for retries to refresh tokens on 401 ([60751a0](https://github.com/whitebox-co/amazon-ads-api/commit/60751a01b202f2e88c7610c1dc90208f111502fa))

# [1.3.0](https://github.com/whitebox-co/amazon-ads-api/compare/v1.2.0...v1.3.0) (2022-03-04)

### Features

-   auto generated updates based on latest amazon schema changes ([9cff834](https://github.com/whitebox-co/amazon-ads-api/commit/9cff8348a3ba73fc3c970849009c22f576b662ac))
-   configuration and adsapi constructor now take a bottleneck limiter as an optional param ([bda0901](https://github.com/whitebox-co/amazon-ads-api/commit/bda09014e04f5c0b025e69f8719a06af4c75d975))

# [1.2.0](https://github.com/whitebox-co/amazon-ads-api/compare/v1.1.0...v1.2.0) (2022-02-01)

### Bug Fixes

-   **deps:** bump axios from 0.21.4 to 0.25.0 ([#53](https://github.com/whitebox-co/amazon-ads-api/issues/53)) ([8f8ff60](https://github.com/whitebox-co/amazon-ads-api/commit/8f8ff60401bbcc4be0d9871fc1a8ec2a1ce4407f))

### Features

-   add throttling and rate-limiting ([#49](https://github.com/whitebox-co/amazon-ads-api/issues/49)) ([5cf5adb](https://github.com/whitebox-co/amazon-ads-api/commit/5cf5adbfe934aa5beb5af8d85b920de8841c4b8a)), closes [#30](https://github.com/whitebox-co/amazon-ads-api/issues/30)
-   auto-generated amazon api updates to brand, data, dsp, and sponsored-products ([6e1313b](https://github.com/whitebox-co/amazon-ads-api/commit/6e1313b8a11286b4d6d17b88e3e15100d6df9084))

# [1.1.0](https://github.com/whitebox-co/amazon-ads-api/compare/v1.0.0...v1.1.0) (2022-01-21)

### Bug Fixes

-   fixed merge conflict typo ([f75d661](https://github.com/whitebox-co/amazon-ads-api/commit/f75d661a91f4d1ac9c02d71f106d847302644e1c))
-   fixed sponsored brand extra overriding sponsored brands resource ([dd8d8ca](https://github.com/whitebox-co/amazon-ads-api/commit/dd8d8cafe3fcfa5142e5370f331a79d1464d6f26))

### Features

-   **attribution:** amazon schema changes ([77e00e4](https://github.com/whitebox-co/amazon-ads-api/commit/77e00e4d02f899cf9c09f8741db06e9f8fcf6376))
-   **sponsored-products-extra:** amazon schema changes ([613793e](https://github.com/whitebox-co/amazon-ads-api/commit/613793efd53d3ed4cc5b1e158906858074941e42))

# 1.0.0 (2022-01-03)

### Bug Fixes

-   adding ability for generator to replace unknown_types so lint does not fail ([dde2762](https://github.com/whitebox-co/amazon-ads-api/commit/dde2762ca42304f128414a8788c5d50ac9e017e2))

### Features


src/helpers.ts
import globalAxios, { AxiosError, AxiosInstance, AxiosResponse } from 'axios';
import Agent from 'agentkeepalive';

import { RequestArgs } from './apis/models/base';
import { Configuration, ConfigurationParameters } from './apis/models/configuration';
import { AmazonAdvertisingAPICredentials, APIConfigurationParameters } from './constants';
import envoyRateLimit, { RateLimitResponseCode } from './connectors/envoy-rate-limit';
import { getConfiguration, getLimiter } from './config';

const httpAgent = new Agent({
	maxSockets: 500,
	timeout: 60 * 1000 * 10, // active socket keepalive for 10 mins
	freeSocketTimeout: 4000, // free socket keepalive for 4 seconds
});

const httpsAgent = new Agent.HttpsAgent({
	maxSockets: 500,
	timeout: 60 * 1000 * 10, // active socket keepalive for 10 mins
	freeSocketTimeout: 4000, // free socket keepalive for 4 seconds
});

/**
 * Adds functions from each individual API class into the main Client Class
 *
 * @param derivedCtor
 * @param baseCtors
 */
export const applyMixins = (derivedCtor: any, baseCtors: any[]): void => {
	for (const baseCtor of baseCtors) {
		for (const name of Object.getOwnPropertyNames(baseCtor.prototype)) {
			const baseCtorName = Object.getOwnPropertyDescriptor(baseCtor.prototype, name);
			if (baseCtorName) {
				Object.defineProperty(derivedCtor.prototype, name, baseCtorName);
			}
		}
	}
};

/**
 * Gets an instance of the global axios singleton.
 *
 * @param {APIConfigurationParameters} parameters
 * @returns {AxiosInstance}
 */
export const getAxiosInstance = (parameters: APIConfigurationParameters): AxiosInstance => {
	let axiosInstance: AxiosInstance;
	const config = getConfiguration();
	const { axios } = parameters;
	const { accessToken, credentials } = parameters;

	const commonHeaders = {
		'Amazon-Advertising-API-ClientId': credentials?.clientId ?? '',
		'Amazon-Advertising-API-Scope': credentials?.profileId ?? '',
		Authorization: `Bearer ${accessToken ?? ''}`,
	};

	if (axios || config.axiosInstance) {
		axiosInstance = config.axiosInstance || axios;
		axiosInstance.defaults.headers.common = commonHeaders;
	} else {
		axiosInstance = globalAxios.create({
			headers: commonHeaders,

			// 10 min timeout
			timeout: 60 * 1000 * 10,

			// keepAlive pools and reuses TCP connections, so it's faster
			httpAgent,
			httpsAgent,

			// follow up to 10 HTTP 3xx redirects
			maxRedirects: 3,
		});
	}

	axiosInstance.interceptors.response.use(
		(response: AxiosResponse) => response,
		(error: AxiosError) => {
			throw error;
		}
	);

	return axiosInstance;
};

/**
 * This used to be part of the `common` OpenAPI generation. However, we needed a custom implementation
 * in order to provide rate-throttling before `axios.request()` is issued.
 *
 * This function is injected via ts-morph into each model after OpenAPI generation is complete.
 */
export const createRequestFunction = (
	axiosArgs: RequestArgs,
	globalAxios: AxiosInstance,
	BASE_PATH: string,
	configuration?: any
) => {
	return async (axios: AxiosInstance = globalAxios, basePath: string = BASE_PATH) => {
		const credentials = configuration?.credentials || {};
		const config = getConfiguration();
		const limiter = getLimiter();
		const axiosRequestArgs = { ...axiosArgs.options, url: (configuration?.basePath || basePath) + axiosArgs.url };
		const endpoint = axiosArgs.url;
		const action = axiosRequestArgs.method;
		const isDspEndpoint = axiosArgs.url.includes('/dsp/');

		// check rate limits against an envoyproxy rate limit service to see if global rate limits are over the allowed // thresholds. - see https://github.com/envoyproxy/ratelimit#json-endpoint for descriptor format
		if (config.envoyProxyRateLimitUrl) {
			const descriptorEntries: any[] = [{ key: 'ads' }];

			if (isDspEndpoint) descriptorEntries.push({ key: 'dsp', value: action });

			const rateLimit = await envoyRateLimit.getLimit('amazon', descriptorEntries);

			if (rateLimit.overallCode === RateLimitResponseCode.OVER_LIMIT) {
				throw new Error(
					`EnvoyRateLimit: Rate Limits for amazon-ads-${endpoint}-${action} are OVER the allowed Limit`
				);
			}
		}

		// wrap the axios request in the limiter if configuration enables throttling
		let request: any = axios.request;
		if (limiter) {
			request = limiter.wrap(axios.request);
			request.withOptions(
				{
					expiration: config.jobOptions?.expiration || 3600 * 1000,
				},
				{ ...axiosRequestArgs, credentials }
			);
		}

		// return the rate limited or normal request.
		return request({ ...axiosRequestArgs, credentials });
	};
};

export interface AdsConfigurationParameters extends ConfigurationParameters {
	credentials?: AmazonAdvertisingAPICredentials;
}

export class AdsConfiguration extends Configuration {
	credentials: AmazonAdvertisingAPICredentials;
	constructor(param: AdsConfigurationParameters = {}) {
		super(param);
		this.credentials = param.credentials;
	}
}
